File: elly/lib/features/emergency/global/domain/services/phone_normalizer.dart
/// phone_normalizer.dart
///
/// International E.164 phone number validator and normalizer.

library;

class PhoneNormalizer {
  /// Map of ISO country codes to international dial calling codes.
  static const Map<String, String> _dialCodes = {
    'IN': '+91',
    'US': '+1',
    'CA': '+1',
    'GB': '+44',
    'AU': '+61',
    'JP': '+81',
    'NZ': '+64',
    'SG': '+65',
    'BR': '+55',
    'MX': '+52',
  };

  /// Normalizes a raw phone string into E.164 standard format.
  static String normalize(String rawPhone, {String defaultCountryCode = 'IN'}) {
    final cleaned = rawPhone.replaceAll(RegExp(r'[^\d+]'), '');

    if (cleaned.isEmpty) return rawPhone;

    if (cleaned.startsWith('+')) {
      return cleaned;
    }

    final dialCode = _dialCodes[defaultCountryCode.toUpperCase()] ?? '+91';

    // Remove leading zeros before prepending dial code
    final trimmedDigits = cleaned.replaceFirst(RegExp(r'^0+'), '');
    return '$dialCode$trimmedDigits';
  }
}

File: backend/app/transcript/phone_normalizer.py
"""
phone_normalizer.py

E.164 international phone number normalizer for Python FastAPI backend.
"""

import re

DIAL_CODES = {
    "IN": "+91",
    "US": "+1",
    "CA": "+1",
    "GB": "+44",
    "AU": "+61",
    "JP": "+81",
    "NZ": "+64",
    "SG": "+65",
    "BR": "+55",
    "MX": "+52"
}


def normalize_phone_number(raw_phone: str, default_country_code: str = "IN") -> str:
    if not raw_phone:
        return raw_phone

    cleaned = re.sub(r"[^\d+]", "", raw_phone)
    if not cleaned:
        return raw_phone

    if cleaned.startswith("+"):
        return cleaned

    dial_code = DIAL_CODES.get(default_country_code.upper(), "+91")
    trimmed_digits = re.sub(r"^0+", "", cleaned)
    return f"{dial_code}{trimmed_digits}"
